Added typescript component types and searchBar change to controlled imput

IlyaSmagin opened the pull request:
### Changes to `SearchBar.tsx` and `Home.tsx` components:
- search field is now controlled input
- lifted search query state and now is is stored in `Home.tsx`, so we don't have to pass `search` function that filtered json into the `SearchBar.tsx` anymore

### Changes to other components:
- defined component types, instead of using `any`

khatabwedaa commented:
Hi @IlyaSmagin, This update made search performance really bad and slow. It will be better if you put that in mind. Thanks again

IlyaSmagin commented:
#### I reverted commit that hurt the performance (it was timing out search input instead of filtering function)

Also added types for initial version `searchBar` component, since it would not affect the performance

## I also looked into ways to remove input timeouts/lag altogether:
- One of the most promising solutions is React 18 `useDeferredValue` hook, as it aims to solve this exact problem, see [docs](https://react.dev/reference/react/useDeferredValue#deferring-re-rendering-for-a-part-of-the-ui)
- But such solution will need updating current version of React this app runs on, and there is compatibility issues with different libraries, but still, I want to explore this option

khatabwedaa commented:
Nice work Thanks :smile:, I can work on the upgrade thing If I have time this weekend
